x-pack/filebeat/input/salesforce/input.go

```diff
@@ -0,0 +1,506 @@
+	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+		s.log.Infof("input stopped because context was cancelled with: %v", err)
+		return nil
+	}
+
+	return err
+}
+
+func (s *salesforceInput) SetupSFClientConnection() (*soql.Resource, error) {
+	if s.sfdcConfig == nil {
+		return nil, errors.New("internal error: salesforce configuration is not set properly")
+	}
+
+	// Open creates a session using the configuration.
+	session, err := session.Open(*s.sfdcConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// set clientSession for re-use (EventLogFile)
+	s.clientSession = session
+
+	// Create a new SOQL resource using the session.
+	soqlr, err := soql.NewResource(session)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up salesforce SOQL resource: %w", err)
+	}
+	return soqlr, nil
+}
+
+// FormQueryWithCursor takes a queryConfig and a cursor and returns a querier.
+func (s *salesforceInput) FormQueryWithCursor(queryConfig *QueryConfig, cursor mapstr.M) (*querier, error) {
+	qr, err := parseCursor(queryConfig, cursor, s.log)
+	if err != nil {
+		return nil, err
+	}
+
+	s.log.Infof("Salesforce query: %s", qr)
+
+	return &querier{Query: qr}, err
+}
+
+// RunObject runs the Object method of the Event Monitoring API to collect events.
+func (s *salesforceInput) RunObject() error {
+	s.log.Debugf("Scrape Objects every %s", s.srcConfig.EventMonitoringMethod.Object.Interval)
+
+	var cursor mapstr.M
+	if !(s.cursor.Object.FirstEventTime == "" && s.cursor.Object.LastEventTime == "") {
+		object := make(mapstr.M)
+		if s.cursor.Object.FirstEventTime != "" {
+			object.Put("first_event_time", s.cursor.Object.FirstEventTime)
+		}
+		if s.cursor.Object.LastEventTime != "" {
+			object.Put("last_event_time", s.cursor.Object.LastEventTime)
+		}
+		cursor = mapstr.M{"object": object}
+	}
+	query, err := s.FormQueryWithCursor(s.config.EventMonitoringMethod.Object.Query, cursor)
+	if err != nil {
+		return fmt.Errorf("error forming based on cursor: %w", err)
+	}
+
+	res, err := s.soqlr.Query(query, false)
+	if err != nil {
+		return err
+	}
+
+	totalEvents := 0
+	firstEvent := true
+
+	for res.TotalSize() > 0 {
+		for _, rec := range res.Records() {
+			val := rec.Record().Fields()
+
+			jsonStrEvent, err := json.Marshal(val)
+			if err != nil {
+				return err
+			}
+
+			if firstEvent {
+				if timstamp, ok := val[s.config.EventMonitoringMethod.Object.Cursor.Field].(string); ok {
+					s.cursor.Object.FirstEventTime = timstamp
+				}
+			}
+
+			if timstamp, ok := val[s.config.EventMonitoringMethod.Object.Cursor.Field].(string); ok {
+				s.cursor.Object.LastEventTime = timstamp
+			}
+
+			err = publishEvent(s.publisher, s.cursor, jsonStrEvent, "Object")
+			if err != nil {
+				return err
+			}
+			firstEvent = false
+			totalEvents++
+		}
+
+		if res.MoreRecords() { // returns true if there are more records.
+			res, err = res.Next()
+			if err != nil {
+				return err
+			}
+		} else {
+			break
+		}
```

Comment:
Another minor improvement and could be simplified as this
```suggestion
		if !res.MoreRecords() { // returns true if there are more records.
			break
		} 
		
		res, err = res.Next()
		if err != nil {
			return err
		}
```

Reply:
Taken care of.